addons/crm/models/crm_lead.py

```diff
@@ -299,14 +299,13 @@ def _compute_company_id(self):
                    (not lead.partner_id or lead.partner_id.company_id != proposal):
                     proposal = False
 
-            # propose a new company based on responsible, limited by team
+            # propose a new company based on team > user (respecting context) > partner
             if not proposal:
-                if lead.user_id and lead.team_id.company_id:
+                if lead.team_id.company_id:
                     proposal = lead.team_id.company_id
                 elif lead.user_id:
-                    proposal = lead.user_id.company_id & self.env.companies
-                elif lead.team_id:
-                    proposal = lead.team_id.company_id
+                    proposal = self.env.company & lead.user_id.company_ids \
```

Comment:
Maybe we could make the & / or condition a bit more verbose but more explicit, like
```
if lead.team_id.company_id:
    proposal = lead.team_id.company_id
elif lead.user_id and self.env.company < lead.user_id.company_ids:
    proposal = self.env.company
elif lead.user_id and lead.user_id.company_id < self.env.companies:
    proposal = lead.user_id.company_id:
elif lead.partner_id:
    proposal = lead.partner_id.company_id
else:
    proposal = False
```